alf/utils/lean_function.py

```diff
@@ -84,16 +140,58 @@ def myfunc(x, w, b, scale=1.0):
 
         y = lean_myfunc(x, w, b)
 
+    2. Apply to nn.Module:
+
+    .. code-block:: python
+
+        module = alf.layers.FC(3, 5, activation=torch.relu_)
+        lean_func = lean_function(module)
+        y = lean_func(x)
+
+    3. Apply to a network
+
+    .. code-block:: python
+
+        net = alf.nn.Sequential(
+            alf.layers.FC(3, 5, activation=torch.relu_),
+            alf.layers.FC(5, 1, activation=torch.sigmoid))
+        lean_func = lean_function(net)
+        y = lean_func(x)
+
     Args:
-        func (Callable): function to be wrapped.
+        func: function or module to be wrapped.
     Returns:
-        the wrapped function.
+        the wrapped function or module. In the case of ``func`` being a ``nn.Module``,
+        all the original attributes and methods can still be accessed in the same
+        way through the wrapped module.
     """
 
-    def wrapped(*args, **kwargs):
+    def _forward(self, *args, **kwargs):
+        if self._inside_lean_function:
+            return self._original_forward_for_lean_function(*args, **kwargs)
+        else:
+            return self._lean_function(*args, **kwargs)
+
+    parameters = ()
+    if isinstance(func, nn.Module):
+        parameters = tuple(func.parameters())
+    if isinstance(func, Network):
+        specs = (func.output_spec, func.state_spec)
+
+    def _wrapped(*args, **kwargs):
         # Function.apply does not allow keyword arguments, so we have to convert
         # all keyword arguments to positional arguments
-        return _LeanFunction.apply(func, tuple(kwargs.keys()), *args,
-                                   *tuple(kwargs.values()))
-
-    return wrapped
+        ret = _LeanFunction.apply(func, len(parameters), tuple(kwargs.keys()),
+                                  *parameters, *args, *tuple(kwargs.values()))
+        if isinstance(func, Network):
+            ret = pack_sequence_as(specs, ret)
+        return ret
+
+    if isinstance(func, nn.Module):
+        func._lean_function = _wrapped
+        func._original_forward_for_lean_function = func.forward
+        func._inside_lean_function = False
+        func.forward = types.MethodType(_forward, func)
```

Comment:
Forgive my ignorance on `Python` - what is the difference here between this and `func.forward = _forward` directly? Is this about adding `_forward` to the method list of `func` as well? 

Is `_forward` also part of pytorch's (hidden) standard API? This could help me understand why we need `_inside_lean_function` too.

Reply:
> Forgive my ignorance on `Python` - what is the difference here between this and `func.forward = _forward` directly? Is this about adding `_forward` to the method list of `func` as well?
> 
> Is `_forward` also part of pytorch's (hidden) standard API? This could help me understand why we need `_inside_lean_function` too.

You may find this helpful:
https://stackoverflow.com/questions/37455426/advantages-of-using-methodtype-in-python

Reply:
Thanks! Coming from C++ I'd thought the method's first argument when called together with a class instance is automatically supplied but apparently it is not so in python. This helps a lot.